### PR TITLE
build: fix the "make O=$(out_dir)" error

### PR DESCRIPTION
**Pull request by prime-zeng**
Signed-off-by: Zeng Tao <prime.zeng@hisilicon.com>

### CODE REVIEW

File: config.mk [hunk 1]
```diff
@@ -40,7 +40,7 @@ CFG_TA_TEST_PATH ?= 1
 
 # Default out dir.
 # Must be a relative path with respect to the op-tee-client root directory
```
Comment (jforissier) [inline config.mk]: Remove this comment, because with your changes it is an absolute path.

File: libsqlfs/Makefile [hunk 2]
```diff
@@ -6,7 +6,7 @@
 include ../flags.mk
 
 O ?= out
-OUT_DIR	?= ${CURDIR}/../$(O)/libsqlfs
+OUT_DIR	?= $(O)/libsqlfs
```
Comment (jforissier) [inline libsqlfs/Makefile]: You're changing the default behavior (when O is not given on the command line). It seems you need to set `O ?= $(CURDIR)/../out`, no?
Same in the 3 other Makefiles below.

File: config.mk (hunk 1 above)
Comment (prime-zeng) [inline config.mk]: OK

File: libsqlfs/Makefile (hunk 2 above)
Comment (prime-zeng) [inline libsqlfs/Makefile]: In fact, what i want to do is that when i use the O option for make, the build binaries will be put into that directory, relative path to the optee-client root directory.

Comment (prime-zeng) [inline libsqlfs/Makefile]: O ?= $(CURDIR)/out is set in the main Makefile and exported, so here is right.

Comment (jforissier) [inline libsqlfs/Makefile]: OK, so remove `O ?= out` above. It is confusing.

**Comment by jenswi-linaro** [comment]
I don't agree that it's more flexible to convert supplied relative paths into absolute paths.
If you don't like relative paths, why use them to start with?

**Comment by jforissier** [comment]
I think the point of this PR is simply to allow absolute paths. With the current code, `make O=/tmp/out` will fail because for `OUT_DIR` in the sub-makefiles we prepend stuff to `$(O)`.

I think what we want is:
```
$ cd /path/to/optee_client
$ make            # Shall create /path/to/optee_client/out and compile to /path/to/optee_client/out/{libteec,libsqlfs,...}
$ make O=foo      # Shall create /path/to/optee_client/foo and compile to /path/to/optee_client/foo/{libteec,libsqlfs,...}
$ make O=/tmp/out # Shall create /tmp/out and write output to /tmp/out/{libteec,libsqlfs,...}
```

The proposed patch does not pass all the above tests, so it's not good.

**Comment by jenswi-linaro** [comment]
OK, misunderstanding by me then. As long as we still support relative paths without conversion into absolute paths I'm OK with it.

**Comment by prime-zeng** [comment]
I have rename the outdir to support make O=/path/to/optee-client, if don't rename, the clean will remove the source. 
but still one problem: make O=~  can't be supported, any suggestion?

**Comment by jforissier** [comment]
@prime-zeng renaming the directories is not the right solution. If the clean rules are broken (I think they are), they need to be fixed. I mean, "rm -rf" should probably never be used in a Makefile ; instead only object files etc. should be removed, and possibly the output directory if empty.

**Comment by prime-zeng** [comment]
@jforissier agree, so we need to fix the clean behavior in the Makefile first, another patch to do this?

**Comment by jforissier** [comment]
@prime-zeng yes, cleaning "make clean" has to be a separate commit. In this PR or in a different one, it doesn't matter.

**Comment by prime-zeng** [comment]
The clean rules has been fixed, but still one problem, make O=~ is not supported, and i use the ifeq($(shell dirname$(O)) to distinguish relative and absolute path, i don't know if  there is a better way to do it.

**Comment by jenswi-linaro** [comment]
Absolute paths starts per definition with `/`. `~` is supposed to be expanded by the shell. In `optee_os/mk/subdir.mk` we test for absolute path in `$(1)` with `ifneq ($(filter /%,$(1)),)`.

**Comment by jforissier** [comment]
Created pull request https://github.com/OP-TEE/optee_client/pull/66. @prime-zeng your comments are welcome, let me know if it works for you.

**Pull request closed without merging**